File: dwl/benchmark/WholeBodyInterface.cpp
#include <dwl/WholeBodyState.h>
#include <dwl/model/WholeBodyKinematics.h>
#include <dwl/model/WholeBodyDynamics.h>
#include <ctime>
#include <chrono>


int main(int argc, char **argv)
{
	// The number of iterations
	unsigned int N = 1000000;

	dwl::WholeBodyState ws;
	dwl::model::FloatingBaseSystem fbs;
	dwl::model::WholeBodyKinematics wkin;
	dwl::model::WholeBodyDynamics wdyn;

	// Resetting the system from the hyq urdf file
	std::string urdf_file = DWL_SOURCE_DIR"/sample/hyq.urdf";
	std::string yarf_file = DWL_SOURCE_DIR"/config/hyq.yarf";
	wdyn.modelFromURDFFile(urdf_file, yarf_file);
	wkin = wdyn.getWholeBodyKinematics();
	fbs = wdyn.getFloatingBaseSystem();

	// Define the DoF after initializing the robot model
	ws.setJointDoF(fbs.getJointDoF());

	wkin.setIKSolver( 1.0e-12, 0.01, 50);


	// The robot state
	ws.setBasePosition_W(Eigen::Vector3d(0., 0., 0.));
	ws.setBaseRPY_W(Eigen::Vector3d(0., 0., 0.));
	ws.setBaseVelocity_W(Eigen::Vector3d(0., 0., 0.));
	ws.setBaseRPYVelocity_W(Eigen::Vector3d(0., 0., 0.));
	ws.setBaseAcceleration_W(Eigen::Vector3d(0., 0., 0.));
	ws.setBaseRPYAcceleration_W(Eigen::Vector3d(0., 0., 0.));
	ws.setJointPosition(0.75, fbs.getJointId("lf_hfe_joint"));
	ws.setJointPosition(-1.5, fbs.getJointId("lf_kfe_joint"));
	ws.setJointPosition(-0.75, fbs.getJointId("lh_hfe_joint"));
	ws.setJointPosition(1.5, fbs.getJointId("lh_kfe_joint"));
	ws.setJointPosition(0.75, fbs.getJointId("rf_hfe_joint"));
	ws.setJointPosition(-1.5, fbs.getJointId("rf_kfe_joint"));
	ws.setJointPosition(-0.75, fbs.getJointId("rh_hfe_joint"));
	ws.setJointPosition(1.5, fbs.getJointId("rh_kfe_joint"));


	dwl::rbd::BodyVector6d grf;
	grf["lf_foot"] << 0, 0, 0, 0, 0, 190.778;
	grf["rf_foot"] << 0, 0, 0, 0, 0, 190.778;
	grf["lh_foot"] << 0, 0, 0, 0, 0, 190.778;
	grf["rh_foot"] << 0, 0, 0, 0, 0, 190.778;


	std::clock_t startcputime = std::clock();
	dwl::rbd::BodyVectorXd contact_pos_W;
	for (unsigned int i = 0; i < N; ++i)
		contact_pos_W = wkin.computePosition(ws.base_pos, ws.joint_pos,
											 fbs.getEndEffectorNames(), // it uses all the end-effector of the system
											 dwl::rbd::Linear, dwl::RollPitchYaw);

	double cpu_duration =
				(std::clock() - startcputime) * 1000000 / (double) CLOCKS_PER_SEC;
	std::cout << "  Forward kinematics: " << cpu_duration / N << " (microsecs, CPU time)" << std::endl;


	dwl::rbd::BodyVector3d ik_pos;
	ik_pos[fbs.getFloatingBaseName()] = Eigen::Vector3d::Zero();
	ik_pos["lf_foot"] = contact_pos_W.find("lf_foot")->second.tail(3);
	ik_pos["rf_foot"] = contact_pos_W.find("rf_foot")->second.tail(3);
	ik_pos["lh_foot"] = contact_pos_W.find("lh_foot")->second.tail(3);
	ik_pos["rh_foot"] = contact_pos_W.find("rh_foot")->second.tail(3);
	dwl::rbd::Vector6d base_pos_init = dwl::rbd::Vector6d::Zero();
	Eigen::VectorXd joint_pos_init(12);
	joint_pos_init << 0., 0.5, -1., 0., -0.5, 1., 0., 0.5, -1., 0., -0.5, 1.;
	startcputime = std::clock();
	for (unsigned int i = 0; i < N; ++i)
		wkin.computeInverseKinematics(ws.base_pos, ws.joint_pos,
									  ik_pos,
									  base_pos_init, joint_pos_init);

	cpu_duration =
				(std::clock() - startcputime) * 1000000 / (double) CLOCKS_PER_SEC;
	std::cout << "  Inverse kinematics: " << cpu_duration / N << " (microsecs, CPU time)" << std::endl;


	startcputime = std::clock();
	for (unsigned int i = 0; i < N; ++i)
		wdyn.computeInverseDynamics(ws.base_eff, ws.joint_eff,
									ws.base_pos, ws.joint_pos,
									ws.base_vel, ws.joint_vel,
									ws.base_acc, ws.joint_acc, grf);

	cpu_duration =
			(std::clock() - startcputime) * 1000000 / (double) CLOCKS_PER_SEC;
	std::cout << "  Inverse dynamics: " << cpu_duration / N << " (microsecs, CPU time)" << std::endl;


	Eigen::MatrixXd inertial_mat;
	startcputime = std::clock();
	for (unsigned int i = 0; i < N; ++i)
		wdyn.computeJointSpaceInertialMatrix(inertial_mat, ws.base_pos, ws.joint_pos);

	cpu_duration =
				(std::clock() - startcputime) * 1000000 / (double) CLOCKS_PER_SEC;
	std::cout << "  Joint space inertia matrix: " << cpu_duration / N << " (microsecs, CPU time)" << std::endl;
	return 0;
}

File: dwl/benchmark/WholeBodyInterface.py
from __future__ import print_function
# This lets us use the python3-style print() function even in python2. It should have no effect if you're already running python3.


import subprocess

# Running the C++ benchmark
print("C++ benchmark:")
args = ("../bin/wif_benchmark")
popen = subprocess.call(args, stdin=None, stdout=None, stderr=None, shell=True)



print("Python benchmark:")
# Running the python benchmark
import dwl
import numpy as np
import time

# Number of iterations
N = 100000

# Construct an instance of the WholeBodyDynamics class, which wraps the C++ class.
ws = dwl.WholeBodyState()
fbs = dwl.FloatingBaseSystem()
wkin = dwl.WholeBodyKinematics()
wdyn = dwl.WholeBodyDynamics()

# Resetting the system from the hyq urdf file
wdyn.modelFromURDFFile("../sample/hyq.urdf", "../config/hyq.yarf")
fbs = wdyn.getFloatingBaseSystem()
wkin = wdyn.getWholeBodyKinematics()


# Define the DoF after initializing the robot model
ws.setJointDoF(fbs.getJointDoF())


# The robot state
ws.setBasePosition_W(np.array([0., 0., 0.]))
ws.setBaseRPY_W(np.array([0., 0., 0.]))
ws.setBaseVelocity_W(np.array([0., 0., 0.]))
ws.setBaseRPYVelocity_W(np.array([0., 0., 0.]))
ws.setBaseAcceleration_W(np.array([0., 0., 0.]))
ws.setBaseRPYAcceleration_W(np.array([0., 0., 0.]))
ws.setJointPosition(0.75, fbs.getJointId("lf_hfe_joint"))
ws.setJointPosition(-1.5, fbs.getJointId("lf_kfe_joint"))
ws.setJointPosition(-0.75, fbs.getJointId("lh_hfe_joint"))
ws.setJointPosition(1.5, fbs.getJointId("lh_kfe_joint"))
ws.setJointPosition(0.75, fbs.getJointId("rf_hfe_joint"))
ws.setJointPosition(-1.5, fbs.getJointId("rf_kfe_joint"))
ws.setJointPosition(-0.75, fbs.getJointId("rh_hfe_joint"))
ws.setJointPosition(1.5, fbs.getJointId("rh_kfe_joint"))

grf = { 'lh_foot' : np.array([0., 0., 0., 0., 0., 190.778]),
        'rf_foot' : np.array([0., 0., 0., 0., 0., 190.778]),
        'lh_foot' : np.array([0., 0., 0., 0., 0., 190.778]),
        'rh_foot' : np.array([0., 0., 0., 0., 0., 190.778]) };
base_eff = ws.base_eff
joint_eff = ws.joint_eff
base_pos = ws.base_pos
joint_pos = ws.joint_pos
base_vel = ws.base_vel
joint_vel = ws.joint_vel
base_acc = ws.base_acc
joint_acc = ws.joint_acc



startcputime = time.clock()
for x in range(0, N):
    contact_pos_B = wkin.computePosition(base_pos, joint_pos,
                                         fbs.getEndEffectorNames(dwl.FOOT),
                                         dwl.Linear)
cpu_duration = (time.clock() - startcputime) * 1000000;
print("  Forward kinematics: ", cpu_duration / N, "(microsecs, CPU time)")



wkin.setIKSolver(1.0e-12, 0.01, 50)
base_pos_init = np.zeros(6);
joint_pos_init = fbs.getDefaultPosture();
startcputime = time.clock()
for x in range(0, N):
    wkin.computeJointPosition(joint_pos, contact_pos_B, joint_pos_init)
cpu_duration = (time.clock() - startcputime) * 1000000;
print("  Inverse kinematics: ", cpu_duration / N, "(microsecs, CPU time)")



startcputime = time.clock()
for x in range(0, N):
    wdyn.computeInverseDynamics(base_eff, joint_eff,
                                base_pos, joint_pos,
                                base_vel, joint_vel,
                                base_acc, joint_acc,
                                grf);
cpu_duration = (time.clock() - startcputime) * 1000000;
print("  Inverse dynamics: ", cpu_duration / N, "(microsecs, CPU time)")



startcputime = time.clock()
joint_inertial_mat = np.zeros([6 + fbs.getJointDoF(), 6 + fbs.getJointDoF()])
for x in range(0, N):
    wdyn.computeJointSpaceInertialMatrix(joint_inertial_mat,
                                         base_pos, joint_pos);
cpu_duration = (time.clock() - startcputime) * 1000000;
print("  Joint space inertia matrix: ", cpu_duration / N, "(microsecs, CPU time)")